File: drills/ts/solutions/nc54_sum_of_two_integers__full_adder.ts
export function getSum(a: number, b: number): number {
  // The same addition written as hardware: thirty-two full adders in a row,
  // each taking two input bits and a carry and producing a sum bit and a carry
  // out. Slower than the XOR loop, which stops as soon as no carries are left,
  // but it is where the XOR loop comes from.
  let result = 0;
  let carry = 0;

  for (let i = 0; i < 32; i++) {
    const x = (a >> i) & 1;
    const y = (b >> i) & 1;
    const xor = x ^ y;
    result |= (xor ^ carry) << i;
    carry = (x & y) | (carry & xor);
  }

  return result;
}

File: drills/go/solutions/nc54_sum_of_two_integers__full_adder.go
package main

func getSum(a int, b int) int {
	// One bit at a time, with an explicit carry, over 32-bit two's complement.
	x, y := uint32(int32(a)), uint32(int32(b))
	var result, carry uint32
	for i := 0; i < 32; i++ {
		bitA := (x >> i) & 1
		bitB := (y >> i) & 1
		sum := bitA ^ bitB ^ carry
		carry = (bitA & bitB) | (bitA & carry) | (bitB & carry)
		result |= sum << i
	}
	return int(int32(result))
}
